prov/gni/src/gnix_tags.c

```diff
@@ -0,0 +1,553 @@
+
+#include "gnix.h"
+#include "gnix_util.h"
+
+struct gnix_tag_storage_ops list_ops;
+struct gnix_tag_storage_ops hlist_ops;
+struct gnix_tag_storage_ops kdtree_ops;
+
+struct gnix_tag_search_element {
+	uint64_t tag;
+	uint64_t ignore;
+	void *context;
+	uint64_t flags;
+	int use_src_addr_matching;
+	struct gnix_address *addr;
+};
+
+/**
+ * @brief converts gnix_tag_list_element to gnix_fab_req
+ *
+ * @param elem  slist element embedded in a gnix_fab_req
+ * @return pointer to gnix_fab_req
+ */
+static inline struct gnix_fab_req *__to_gnix_fab_req(
+		struct gnix_tag_list_element *elem)
+{
+	struct gnix_fab_req *req;
+
+	req = container_of(elem, struct gnix_fab_req, tle);
+
+	return req;
+}
+
+/**
+ * @brief determines if a req matches the address parameters
+ *
+ * @param req         gnix fabric request
+ * @param addr        address to match
+ * @param addr_ignore bits in address to ignore
+ * @return 0 if the request does not match the parameters, 1 otherwise
+ */
+static inline int __req_matches_addr_params(
+		struct gnix_fab_req *req,
+		struct gnix_address *addr)
+{
+	return (GNIX_ADDR_UNSPEC(*addr) ||
+			GNIX_ADDR_EQUAL(req->addr, *addr));
+}
+
+int _gnix_req_matches_params(
+		struct gnix_fab_req *req,
+		uint64_t tag,
+		uint64_t ignore,
+		uint64_t flags,
+		void *context,
+		int use_src_addr_matching,
+		struct gnix_address *addr)
+{
+	int valid_request;
+
+	if ((flags & FI_CLAIM) && (flags & FI_PEEK))
+		valid_request = !req->tle.claimed && context != NULL;
+	else if ((flags & FI_CLAIM) && !(flags & FI_PEEK))
+		valid_request = req->tle.claimed && context == req->tle.context;
+	else
+		valid_request = !req->tle.claimed;
+
+	if (use_src_addr_matching)
+		valid_request &= __req_matches_addr_params(req, addr);
+
+	return ((req->tag & ~ignore) == (tag & ~ignore)) && valid_request;
+}
+
+/* used to match elements in the posted lists */
+int _gnix_match_posted_tag(struct slist_entry *entry, const void *arg)
+{
+	const struct gnix_tag_search_element *s_elem = arg;
+	struct gnix_tag_list_element *tle;
+	struct gnix_fab_req *req;
+
+	tle = (struct gnix_tag_list_element *) entry;
+	req = __to_gnix_fab_req(tle);
+
+	return _gnix_req_matches_params(req, s_elem->tag, req->ignore_bits,
+			s_elem->flags, s_elem->context,
+			s_elem->use_src_addr_matching,
+			s_elem->addr);
+}
+
+/* used to match elements in the unexpected lists */
+int _gnix_match_unexpected_tag(struct slist_entry *entry, const void *arg)
+{
+	const struct gnix_tag_search_element *s_elem = arg;
+	struct gnix_tag_list_element *tle;
+	struct gnix_fab_req *req;
+
+	tle = (struct gnix_tag_list_element *) entry;
+	req = __to_gnix_fab_req(tle);
+
+	return _gnix_req_matches_params(req, s_elem->tag, s_elem->ignore,
+			s_elem->flags, s_elem->context,
+			s_elem->use_src_addr_matching,
+			s_elem->addr);
+}
+
+/* default attributes for tag storage objects */
+static struct gnix_tag_storage_attr default_attr = {
+		.type = GNIX_TAG_AUTOSELECT,
+		.use_src_addr_matching = 0,
```

Comment:
When Howard discussed the original idea for including address matching, he mentioned that the address matching should be something of an option and not in the common code path unless requested. Disabling address matching by default follows that line of thinking